packages/core/src/validate/run/initializer.ts

```diff
@@ -318,22 +321,55 @@ function getRecursiveFunctionIds(
 }
 
 /**
- * Get all functions that could be initializers. Does not include private functions.
- * For parent contracts, only internal and public functions which contain statements are included.
+ * Get all functions that are annotated as initializers or are inferred to be initializers.
+ * Logs a note if any reinitializer is found.
  */
-function getPossibleInitializers(contractDef: ContractDefinition, isParentContract: boolean) {
+function getPossibleInitializers(
+  contractDef: ContractDefinition,
+  isParentContract: boolean,
+  decodeSrc: SrcDecoder,
+): FunctionDefinition[] {
   const fns = [...findAll('FunctionDefinition', contractDef)];
-  return fns.filter(
-    fnDef =>
-      (fnDef.modifiers.some(modifier => ['initializer', 'onlyInitializing'].includes(modifier.modifierName.name)) ||
-        ['initialize', 'initializer'].includes(fnDef.name)) &&
-      // Skip virtual functions without a body, since that indicates an abstract function and is not itself an initializer
-      !(fnDef.virtual && !fnDef.body) &&
-      // Ignore private functions, since they cannot be called outside the contract
-      fnDef.visibility !== 'private' &&
-      // For parent contracts, only internal and public functions which contain statements need to be called
-      (isParentContract
-        ? fnDef.body?.statements?.length && (fnDef.visibility === 'internal' || fnDef.visibility === 'public')
-        : true),
+  return fns.filter((fnDef: FunctionDefinition) => {
+    const validateAsInitializer = hasValidateAsInitializerAnnotation(fnDef, decodeSrc);
+    if (!validateAsInitializer && fnDef.modifiers.some(modifier => 'reinitializer' === modifier.modifierName.name)) {
+      logNote(`Reinitializers are not included in validations by default`, [
+        `${decodeSrc(fnDef)}: If you want to validate this function as an initializer, annotate it with '@custom:oz-upgrades-validate-as-initializer'`,
+      ]);
+    }
+
+    return validateAsInitializer || inferPossibleInitializer(fnDef, isParentContract);
+  });
+}
+
+function hasValidateAsInitializerAnnotation(node: Node, decodeSrc: SrcDecoder): boolean {
+  const doc = getDocumentation(node);
+  const tag = 'oz-upgrades-validate-as-initializer';
+  const validateAsInitializer = hasAnnotationTag(doc, tag);
+  if (validateAsInitializer) {
+    const annotationArgs = getAnnotationArgs(doc, tag);
+    if (annotationArgs.length !== 0) {
+      throw new Error(`${decodeSrc(node)}: @custom:${tag} annotation must not have any arguments`);
+    }
+  }
+  return validateAsInitializer;
+}
+
+/**
+ * Infers whether a function could be an initializer. Does not include private functions.
+ * For parent contracts, only internal and public functions which contain statements are included.
```

Comment:
I would expand on why internal functions can be considered a valid initializer. I would expect the initializer is by definition, an external function called into the contract to set it up.

Reply:
There are two different things that we are referring to as initializers.  For the child contract, this is the external function that should be called to set up.  But for parent contracts, these can be internal functions like `__ERC20_init()` that need to be inherited and called by the child initializer.

I think this new annotation needs some rework to account for this, for example if the annotation is on a private function on a parent contract, that should not be allowed.  Otherwise no child contract can call it.

Reply:
In Line 369 we're filtering private functions, so those are already not considered, right?

I get the point of running validation in internal functions as in the case o `__ERC20_init()`

Reply:
That logic was previously short circuited on line 341 if the annotation was present.  Changed it to ignore those cases (e.g. private) even if the annotation is present, and added documentation to clarify this behavior.

Reply:
Right, I see the annotation was preceding those conditions. Solving